### PR TITLE
Correct handling of StgDict/StgInfo for Structure return types

### LINKED ISSUE #444: ctypes_patch breaks with Python 3.13.0a6
### Describe the bug

As of Python3.13.0a6, the ctypes patch used to allow struct return values no longer works.

### Steps to reproduce

Run the test suite on Python 3.13.0a6. 

There will be 7 test failures, all of the form:
```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/unittest/case.py", line 651, in run
    self._callTestMethod(testMethod)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/unittest/case.py", line 606, in _callTestMethod
    if method() is not None:
       ~~~~~~^^
  File "/Users/runner/work/rubicon-objc/rubicon-objc/tests/test_ctypes_patch.py", line 96, in test_patch_idempotent
    ctypes_patch.make_callback_returnable(TestStruct)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/Users/runner/work/rubicon-objc/rubicon-objc/.tox/py/lib/python3.13/site-packages/rubicon/objc/ctypes_patch.py", line 197, in make_callback_returnable
    stgdict = get_stgdict_of_type(ctype)
              ~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/Users/runner/work/rubicon-objc/rubicon-objc/.tox/py/lib/python3.13/site-packages/rubicon/objc/ctypes_patch.py", line 175, in get_stgdict_of_type
    raise TypeError(
        f"The given type's dict must be a StgDict, not {type(stgdict).__module__}.{type(stgdict).__qualname__}"
    )
TypeError: The given type's dict must be a StgDict, not builtins.dict
```

### Expected behavior

Test suite should pass.

### Screenshots

_No response_

### Environment

- Operating System: macOS
- Python version: 3.13.0a6

### Logs

_No response_

### Additional context

Test suite passes with 3.13.0a5.

### PR DESCRIPTION
**Pull request by freakboy3742**
Fixes #444.

python/cpython#114314 replaced the StgDict structure with StgInfo; this has been released in Python 3.13.0a6. 

The real fix for this is to address this upstream, rather than requiring us to work around ctypes; but:

1. Working up this patch was helpful in understanding how the new StgInfo structure works
2. Until such time as a patch is merged upstream, we need a workaround so that Rubicon's tests pass.

Although the patch looks like a lot of code, most of the delta is actually a re-org so that the older code will be auto removed when Py3.13 is the oldest supported version.

As a test, I've also run the full Toga testbed on macOS with this code and it worked as expected.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: src/rubicon/objc/ctypes_patch.py [hunk 1]
```diff
@@ -115,68 +89,133 @@ class ffi_type(ctypes.Structure):
...
+        # type data for that type; it can be found by starting at the memory
+        # address of the type, and offsetting by the tp_basicsize offset of the
+        # base of the CType_Type class - which is `type`.
+        py_type_type = PyTypeObject.from_address(id(type))
 
-    if not isinstance(tp, type):
-        raise TypeError(
-            f"Expected a type object, not {type(tp).__module__}.{type(tp).__qualname__}"
-        )
+        return StgInfo.from_address(id(tp) + py_type_type.tp_basicsize)
```
[the remark this answers is not in the document]
Comment (dgelessus) [reply src/rubicon/objc/ctypes_patch.py:214]: After obtaining the `StgInfo`, we should also check that its `initialized` field is non-zero, like the real code in CPython does.

**Review by dgelessus — CHANGES_REQUESTED** [review]
Thank you for taking care of this!

**Comment by mhsmith** [comment]
I'm not familiar with this area, so I'll leave the reviewing to others.

**Review by samschott — COMMENTED** [review]
Thanks for the fix, and the very well documented code! Pending the requested changes by @dgelessus, this looks good to me.

**Review by samschott — CHANGES_REQUESTED** [review]

File: src/rubicon/objc/ctypes_patch.py [hunk 2]
```diff
@@ -115,68 +89,145 @@ class ffi_type(ctypes.Structure):
...
 
-    if not isinstance(tp, type):
-        raise TypeError(
-            f"Expected a type object, not {type(tp).__module__}.{type(tp).__qualname__}"
+        # tp is the Python representation of the type. The StgInfo struct is the
+        # type data stored on ctypes.CType_Type (which is the base class of
+        # ctypes.Structure).
+        info = ctypes.pythonapi.PyObject_GetTypeData(
+            id(tp),
+            id(type(ctypes.Structure).__base__),
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply src/rubicon/objc/ctypes_patch.py:222]: The internal call that we're replicating is:

    PyObject_GetTypeData((PyObject *)type, state->PyCType_Type)

`type` is the base Python type, which has different handling. `type(tp)` is `_ctypes.PyCStructType`, which is a subclass of `_ctypes.PyCType_Type`, and has it's own type data.  `type(tp).__base__` resolves to `PyCType_Type` for a simple case, but if there's an inheritance chain in place, that won't be guaranteed; using `type(ctypes.Structure).__base__` guarantees we always get `PyCType_Type`.

I guess the real fix here is that the initial check on L207 isn't consistent with the original method - it should be checking `state->PyCType_Type` as well. I'll make that change, and added in references to the original code as well.

(The other answer, of course, is that using `id(type(tp))` or `id(type)` doesn't work at runtime, but `id(type(ctypes.Structure).__base__)` does 😄)

[the remark this answers is not in the document]
Comment (samschott) [reply src/rubicon/objc/ctypes_patch.py:222]: Thanks, and thanks for the explanation!

Btw, `id(type)` works just fine for me at runtime and `id(type(ctypes.Structure).__base__) == id(type)` evaluates to True.

**Review by samschott — APPROVED** [review]

File: src/rubicon/objc/ctypes_patch.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply src/rubicon/objc/ctypes_patch.py:222]: Are you running your test on 3.13.0a6? `id(type(ctypes.Structure).__base__) == id(type)` definitely returns True on Python 3.10, 3.12, and presumably everything before 3.13.0a6, but it's False on 3.13.0a6; and if you're not on 3.13.0a6, the GetTypeData call won't be invoked at all.

[the remark this answers is not in the document]
Comment (samschott) [reply src/rubicon/objc/ctypes_patch.py:222]: Ohh, that explains it!

**Comment by freakboy3742** [comment]
@dgelessus Are you OK with this updated version?

**Review by dgelessus — APPROVED** [review]
LGTM as well. Thanks!

**Pull request merged**